### PR TITLE
fix(coverage): treat assert/require as lines instead branch

### LINKED ISSUE #4294: feat(forge): disable coverage for assertions
### Component

Forge

### Describe the feature you would like

Assertions are never meant to fail unless there is a bug, but Forge currently reports partial coverage for `assert` statements:

Partial coverage for assertions

Related issue in the `solidity-coverage` repo: https://github.com/sc-forks/solidity-coverage/issues/269.

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
Closes #4294 

- assert/require are treated as branches hence always reporting less coverage (branch 0 not hit)

## Solution

- remove assert/require check (do not add branches but treat them as regular lines instead, already added as line kind when `self.push_item` as a statement)
- remove duped code, group function node type with others
- unit test to make sure assert/require does not add branches

### CODE REVIEW

File: crates/evm/coverage/src/analysis.rs [hunk 1]
```diff
@@ -322,33 +325,6 @@ impl<'a> ContractVisitor<'a> {
...
-
-                let expr: Option<Node> = node.attribute("expression");
-                if let Some(NodeType::Identifier) = expr.as_ref().map(|expr| &expr.node_type) {
-                    // Might be a require/assert call
-                    let name: Option<String> = expr.and_then(|expr| expr.attribute("name"));
-                    if let Some("assert" | "require") = name.as_deref() {
-                        self.push_branches(&node.src, self.branch_id);
-                        self.branch_id += 1;
-                    }
-                }
```
Comment (mattsse) [inline crates/evm/coverage/src/analysis.rs:340]: ah okay, this is now no longer required because all this did was check for assert | require

**Review by mattsse — APPROVED** [review]
great!

File: crates/forge/tests/cli/coverage.rs [hunk 2]
```diff
@@ -188,3 +188,53 @@ contract BContractTest is DSTest {
...
+        lcov_info.to_str().unwrap().to_string(),
+    ]);
+    cmd.assert_success();
+    assert!(lcov_info.exists());
+
+    let lcov_data = std::fs::read_to_string(lcov_info).unwrap();
+    // Should find no branch for assert / require, BRF:0 (BRF: branches found).
+    let re = Regex::new(r"BRF:0").unwrap();
+    let valid_line = |line| re.is_match(line);
+    assert!(lcov_data.lines().any(valid_line), "{lcov_data}");
```
Comment (DaniPopes) [inline crates/forge/tests/cli/coverage.rs:239]: can you please just use .assert() on the command: #8406

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/forge/tests/cli/coverage.rs (hunk 2 above)
Comment (grandizzy) [reply crates/forge/tests/cli/coverage.rs]: pls check https://github.com/foundry-rs/foundry/pull/8413/commits/7eb98471f63988241381701ddec09012fd7bc613 if OK I can port couple of more coverage tests to this approach in same PR

Comment (grandizzy) [reply crates/forge/tests/cli/coverage.rs]: oops, pushed porting for other coverage tests by mistake with https://github.com/foundry-rs/foundry/pull/8413/commits/8b1e2ad9866bbed271c675be05676e985577a0ce :)

Comment (grandizzy) [reply crates/forge/tests/cli/coverage.rs]: I think I got it right now, thanks! pls recheck

**Review by DaniPopes — APPROVED** [review]

**Comment by KholdStare** [comment]
@mattsse @grandizzy @DaniPopes I think `require` should still be a branch. Can you check my reasoning? 

There is a subtle difference between `require` and `assert`:
 - `assert` deals with asserting invariants that should hold no matter what the input is. Hence why it shouldn't be possible to hit "the other branch" - that would indicate a bug in the code.
 - `require` deals with pre/post conditions that may get violated due to erroneous/malicious inputs. Both sides of the branch need to be checked as both cases can occur. That's why we have `expectRevert` in tests.

There is a semantic difference and it's very well highlighted [in the solidity docs](https://docs.soliditylang.org/en/latest/control-structures.html#panic-via-assert-and-error-via-require):

> Assert should only be used to test for internal errors, and to check invariants. Properly functioning code should never create a Panic, not even on invalid external input. If this happens, then there is a bug in your contract which you should fix. Language analysis tools can evaluate your contract to identify the conditions and function calls which will cause a Panic.

There is also a concrete/practical difference - [the `SMTChecker` in `solc` relies on this distinction](https://docs.soliditylang.org/en/latest/smtchecker.html):

> he SMTChecker module automatically tries to prove that the code satisfies the specification given by require and assert statements. That is, it considers require statements as assumptions and tries to prove that the conditions inside assert statements are always true. If an assertion failure is found, a counterexample may be given to the user showing how the assertion can be violated.

Does this make sense?

**Pull request merged**